DragaliaAPI/Services/SummonService.cs

```diff
@@ -10,21 +19,94 @@ public class SummonService : ISummonService
     private readonly IUnitDataService _unitDataService;
     private readonly IDragonDataService _dragonDataService;
 
+    private const float SSRSummonRateChara = 0.5f;
+    private const float SSRSummonRateDragon = 0.8f;
+    private const float SRSummonRateTotalNormal = 9.0f;
+    private const float SRSummonRateTotalFeatured = 7.0f;
+    private const float SRSummonRateTotal = SRSummonRateTotalNormal + SRSummonRateTotalFeatured;
+    private const float RSummonRateChara = 4.0f;
+
     public SummonService(IUnitDataService unitDataService, IDragonDataService dragonDataService)
     {
         _unitDataService = unitDataService;
         _dragonDataService = dragonDataService;
     }
 
-    public List<SummonEntity> GenerateSummonResult(int numSummons)
+    public record BannerSummonInfo(
+        Dictionary<EntityTypes, SummonableEntity> featured,
+        Dictionary<EntityTypes, SummonableEntity> normal,
+        double baseSsrRate,
+        double baseRRate
+    );
```

Comment:
I tend to prefer declaring types in their own file rather than in services/controllers. I guess it just makes them easier to find?

Reply:
I was working on figuring out proper rolling when we decided to put it as polish, this was port of it. It is only used in the unused method for figuring out odds and probably should eventually it should go into its own file when it contains everything the method needs to calculate them